### PR TITLE
BugFix: Parallel Data Processing

### PR DESCRIPTION
**Pull request by cyantangerine**
Parallel Data Processing Error BugFix

## Description
Removed the Parallel(.., ,return_as="generator_unordered") method in sdgx/models/components/optimize/sdv_ctgan/data_transformer.py, instead of Parallel(n_jobs=-1, return_as="generator").

## Motivation and Context
In the data preprocessing of the CTGAN model, the DataTransformer parallelizes the DataFrame data column by column, and after processing, assembles the columns to generate a 2D tensor. Due to the lack of consideration for the order returned by parallelization, the columns of the DataFrame and NDArray may not correspond, resulting in the training data not matching the target column correctly, which in turn affects the quality of the model and generated data.
This bug can be recurrence by fitting a dataset larger than 1k records which has crossing order of categorical and continuous variable columns. When we are fitting it, numpy will report a warning of caculating a logarithm of negative. But we expect the number to log is a sum of one-hot array (frequency of categorical), it never be negative. By researching, this bug caused it by changing the trainning data columns order unmatching the original tables and features.

## How has this been tested?
By restarting the fitting after the fix, the warning is disappeared, which means the logarithm of frequency has been caculated correctly.

## Types of changes

- [ ] Maintenance (no change in code, maintain the project's CI, docs, etc.)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by jalr4ever** [comment]
@cyantangerine Please check the review and then proceed with the testing in the current way.

**Comment by cyantangerine** [comment]
Firstly, I tried testing in `pytest` and normal CLI, but it is hard to recurrent the phenomenon —— showing the warning. While, using same dataset in Jupyter Notebook, the warning has been found in every trail. By reading the documents of `Parallel`, it says:
> If 'generator_unordered', the generator will immediately yield
            available results independently of the submission order. The output
            order is not deterministic in this case because it depends on the
            concurrency of the workers.

It can answer why the bug had occerence, so I tried another way: Find columns which was transformed by OneHotEncoder but not filled with 0 and 1. While, using the old method to transform we can see some continuous columns has been switched to discrete columns in the screenshot. It has been confirmed the switching of columns which described in a new commit. What's more, why the `Notebook` environment always lead to this warning and CLI not, i am not understanding right now.

截图

So, it better to change the Parallel return method for more compatibility. And, as the document said, current method is questioned even if it not raised exception this time, but we can not guarantee it in various hardware.

File: sdgx/models/components/optimize/sdv_ctgan/data_transformer.py [hunk 1]
```diff
@@ -170,12 +170,15 @@ def _parallel_transform(self, raw_data, column_transform_info_list) -> NDArrayLo
...
                 process = delayed(self._transform_discrete)(column_transform_info, data)
             processes.append(process)
-        try:
-            # For Future versions of joblib
-            p = Parallel(n_jobs=-1, return_as="generator_unordered")
-            logger.warning("Using generator_unordered in joblib, time to remove this exception!")
-        except ValueError:
-            p = Parallel(n_jobs=-1, return_as="generator")
+
+        # TODO This method has bug, which are deprecated. Time to find a better method instead.
```
Comment (jalr4ever) [inline sdgx/models/components/optimize/sdv_ctgan/data_transformer.py:174]: I think the code comments here are redundant and can be removed.

File: tests/optmize/data_for_test_parallel_transform.csv
Comment (jalr4ever) [inline tests/optmize/data_for_test_parallel_transform.csv:1]: I think we should not add test data like CSV into the code, as it will become larger with project updates. 

Here, you can describe the schema you want using code and generate a DataFrame in memory for testing. As of now, generating 5,000 records using `benchmarks/generate_dataset.py` in the repository seems to trigger the phenomenon you described. 

Of course, this is just for reference. Our focus is on **not adding test data-files to the repository** ; you can generate CSV data according to your own code.

File: tests/optmize/test_parallel_transform.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    data_connector = CsvConnector(path=dataset_csv)
+    data_loader = DataLoader(data_connector)
+    data_metadata = Metadata.from_dataloader(data_loader)
+    data_metadata.datetime_format = {key: "%Y/%m/%d" for key in data_metadata.datetime_columns}
+    transformer = DataTransformer()
+    transformer.fit(data_loader, data_metadata.discrete_columns)
+    return transformer, data_loader
+
+
+# def unfixed_parallel_transform(self, raw_data, column_transform_info_list):
```
Comment (jalr4ever) [inline tests/optmize/test_parallel_transform.py:33]: The redundant code here can be deleted.

File: tests/optmize/test_parallel_transform.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    # start test
+    ndarry_loader = transformer._parallel_transform(
+        data_loader, transformer._column_transform_info_list
+    )
+    data_sampler = DataSampler(ndarry_loader, transformer.output_info_list, True)
+
+
+# def test_parallel_transform_unfixed():
+#     transformer, data_loader = preparing_data()
+#     # start test
```
Comment (jalr4ever) [inline tests/optmize/test_parallel_transform.py:63]: The redundant code here can be deleted.

File: tests/optmize/test_parallel_transform.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    ndarry_loader = transformer._parallel_transform(
+        data_loader, transformer._column_transform_info_list
+    )
+
+    find_not_matching_column_type(ndarry_loader, transformer._column_transform_info_list)
+
+
+# def test_parallel_transform_unfixed_caused_columns_switching():
+#     unfixed_transformer, unfixed_data_loader = preparing_data()
+#     unfixed_ndarry_loader = unfixed_parallel_transform(unfixed_transformer, unfixed_data_loader,
```
Comment (jalr4ever) [inline tests/optmize/test_parallel_transform.py:94]: The redundant code here can be deleted.

File: tests/optmize/test_parallel_transform.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+#     # start test
+#     ndarry_loader = unfixed_parallel_transform(transformer, data_loader, transformer._column_transform_info_list)
+#     data_sampler = DataSampler(ndarry_loader, transformer.output_info_list, True)
+
+
+def find_not_matching_column_type(data, column_transform_info_list):
+    col_index = 0
+    for column_transform_info in column_transform_info_list:
+        output_dim = column_transform_info.output_dimensions
+        if column_transform_info.column_type == "discrete":
```
Comment (jalr4ever) [inline tests/optmize/test_parallel_transform.py:72]: This judgment is incorrect; **it needs to be supplemented that the encoder type is one-hot**. Because for discrete data, label encoding or target encoding does not represent this kind of 0 and 1 data.

File: sdgx/models/components/optimize/sdv_ctgan/data_transformer.py [hunk 6]
```diff
@@ -170,12 +170,14 @@ def _parallel_transform(self, raw_data, column_transform_info_list) -> NDArrayLo
...
-        try:
-            # For Future versions of joblib
-            p = Parallel(n_jobs=-1, return_as="generator_unordered")
-            logger.warning("Using generator_unordered in joblib, time to remove this exception!")
-        except ValueError:
-            p = Parallel(n_jobs=-1, return_as="generator")
+
+        # try:
+        #     # For Future versions of joblib
+        #     p = Parallel(n_jobs=-1, return_as="generator_unordered")
```
Comment (jalr4ever) [inline sdgx/models/components/optimize/sdv_ctgan/data_transformer.py:176]: There is still redundant commented-out code here that needs to be deleted.

**Review by jalr4ever — COMMENTED** [review]
Approved.

**Comment by cyantangerine** [comment]
> Approved.

You may need to select a comment type (for Approved) in File Changes tab.  : )

**Review by jalr4ever — APPROVED** [review]
Approved.

**Comment by Wh1isper** [comment]
Seems a bug fix, @jalr4ever would you like to make a release for this?

**Comment by jalr4ever** [comment]
@Wh1isper Yeah, I plan to do a release of the recent work, it shall be next week.
